metricbeat/module/kafka/consumergroup/query.go

```diff
@@ -113,12 +114,22 @@ func fetchGroupInfo(
 
 		for topic, partitions := range ret.off.Blocks {
 			for partition, info := range partitions {
+				partitionOffset, err := getPartitionOffsetFromTheLeader(b, topic, partition)
```

Comment:
Something to explore here.

I guess that it may happen that the partition offset here is always going to be ahead of the group offset, because we get first the group offset, and then the partition offset. Between both operations the partition offset may have changed.

Starting on version 4 of [`ListOffsets`](https://kafka.apache.org/protocol#The_Messages_ListOffsets) (the API method used to get partition offsets), it is possible to indicate a `current_leader_epoch` to retrieve "old" metadata.

Starting on version 5 of [`OffsetFetch`](https://kafka.apache.org/protocol#The_Messages_OffsetFetch) (the API method used to get consumer group offsets), its response contains a `leader_epoch` field.

I wonder if we can use the `leader_epoch` contained in the response of `OffsetFetch` when available to query for the offset of the partition in the same epoch. This way we could have a more accurate value.

Reply:
What are the implications for the support matrix? I'm not very familiar with API version vs Kafka versioning.

Reply:
@jsoriano not sure if this can be achieved with the current implementation of (GetOffset)[https://github.com/elastic/beats/blob/f35ffdca261d49c83a2a8fcbd40b2cb107c95b43/vendor/github.com/Shopify/sarama/client.go#L62]

Reply:
> What are the implications for the support matrix? I'm not very familiar with API version vs Kafka versioning.

All messages in kafka protocol are versioned, each client and broker can support a different range of versions for each message. There is a message (`ApiVersionsRequest`) to query the versions supported by the broker, we could use this method to decide if we can use the methods aware of the epoch.

> @jsoriano not sure if this can be achieved with the current implementation of [GetOffset](https://github.com/elastic/beats/blob/f35ffdca261d49c83a2a8fcbd40b2cb107c95b43/vendor/github.com/Shopify/sarama/client.go#L62)

No, we would need to forge our own request as we do to [request partition offsets to the leader](https://github.com/elastic/beats/blob/68ce4520a46335cdec890800b0004d4cf611e5c3/metricbeat/module/kafka/broker.go#L166). Or we could contribute to Sarama the support for these versions. :slightly_smiling_face:

Reply:
👍 +1